### DISCUSSION TITLE
Can't connect to Mosquitto TLS

### DISCUSSION THREAD

**Discussion by smilexth** [arendst/Tasmota#11378] [Q&A] [answered]
### PROBLEM DESCRIPTION

1. I have Mosquitto installed with TLS enabled on Docker container (https://github.com/PythonLinks/letsencrypt-mosquitto)
2. Let's Encrypt looks fine
3. Everything works fine when i connected to 1883 but when try to connect to 8883 it has an error with **code : 296** which describted as **"SSL3_ALERT_HANDSHAKE_FAILURE"** according to TLS common error at https://tasmota.github.io/docs/TLS/
4. I have **enabled tls on tasmota self-compiled** with suggested parameters as describted at https://tasmota.github.io/docs/TLS/ with auto learning param

### REQUESTED INFORMATION

- [X] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [X] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [X] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [X] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [X] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [X] Device used (e.g., Sonoff Basic): _____
- [X] Tasmota binary firmware version number used: 9.3.1
  - [ ] Pre-compiled
  - [X] Self-compiled

### TO REPRODUCE

1. Everything seem fine when switch back to connecting via port 1883 with same mqtt credentials but error still remaining when switch to port 8883 with same credential provided.
2. I tried to define  **#define MQTT_FINGERPRINT1 with auto-learning (It won't fix)**
3. I tried to define **#define USE_MQTT_TLS_CA_CERT  (It won't fix)**

### EXPECTED BEHAVIOUR

Successfully connected via port 8883

### SCREENSHOTS

**Mosquitto Startup Logs**

![dddasdasdasdasdas](https://user-images.githubusercontent.com/154602/111458098-9f4f3f80-874b-11eb-9d86-2cc7b990cc64.png)

**Mosquitto Error Log when try to connecting to 8883**

![Untitled](https://user-images.githubusercontent.com/154602/111458781-6f546c00-874c-11eb-8bd4-758a03036a2d.png)

**user_config_override.h**

![Untitled](https://user-images.githubusercontent.com/154602/111458864-914dee80-874c-11eb-8050-b129e446d429.png)

**Comment by Jason2866** [comment]
You are using a LetsEncrypt Mosquitto build. You have to use `#define USE_MQTT_TLS_CA_CERT`
Have you tested your mqtt broker setup?
Take a look https://medium.com/@shayaman321/how-to-setup-a-secured-mqtt-broker-c43d4acb95bf

**Comment by smilexth** [comment]
> You are using a LetsEncrypt Mosquitto build. You have to use `#define USE_MQTT_TLS_CA_CERT`
> Have you tested your mqtt broker setup?
> Take a look https://medium.com/@shayaman321/how-to-setup-a-secured-mqtt-broker-c43d4acb95bf

I tried **#define USE_MQTT_TLS_CA_CERT** but still get no luck

**Comment by smilexth** [comment]
with **#define USE_MQTT_TLS_CA_CERT** got an error from mqtt server side as **"no shared cipher"**

any idea ?

**Comment by smilexth** [comment]
Now i tried to test with mosquitto_sub to 8883, All works fine. but still can't connect from tasmota

**Comment by s-hadinger** [comment]
Can you give details about your mosqitto tls configuration ? Hostname ? Certificate?

**Comment by Elix-g** [comment]
Also interesting: TLS ciphers that are available on server site. You can find out which ciphers your server supports using nmap and/or openssl:

TLSv1.2:
`nmap --script ssl-enum-ciphers -p 8883 mqtt.server.hostname` or:
`openssl s_client -tls1_2 -connect mqtt.server.hostname:8883`

TLSv1.3:
`openssl s_client -tls1_3 -connect mqtt.server.hostname:8883`

I'm using mosquitto as central mqtt server with TLS (v1.2) on port 8883 without any issues.

**Comment by smilexth** [comment]
> Also interesting: TLS ciphers that are available on server site. You can find out which ciphers your server supports using nmap and/or openssl:
> 
> TLSv1.2:
> `nmap --script ssl-enum-ciphers -p 8883 mqtt.server.hostname` or:
> `openssl s_client -tls1_2 -connect mqtt.server.hostname:8883`
> 
> TLSv1.3:
> `openssl s_client -tls1_3 -connect mqtt.server.hostname:8883`
> 
> I'm using mosquitto as central mqtt server with TLS (v1.2) on port 8883 without any issues.

```
No client certificate CA names sent
Peer signing digest: SHA256
Peer signature type: ECDSA
Server Temp Key: X25519, 253 bits
---
SSL handshake has read 2703 bytes and written 304 bytes
Verification: OK
---
New, TLSv1.2, Cipher is ECDHE-ECDSA-AES256-GCM-SHA384
Server public key is 256 bit
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : ECDHE-ECDSA-AES256-GCM-SHA384
    Session-ID: D3F4D18F93F684613C66C85EA48AA41F492349377189F8A6EECF996B27D79D48
    Session-ID-ctx:
    Master-Key: 91FB27B0EA2F924A34EF93CF7CD4A679EF0ED45FEC6ED9BB5F6E6701067A0D238931633C890088B764935CD12F9A17B1
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    TLS session ticket lifetime hint: 7200 (seconds)
    TLS session ticket:
    0000 - ff ac 12 68 1f ea 74 01-0b 2e ca 3a 6c ec 85 db   ...h..t....:l...
    0010 - a6 a5 cd f8 6e 55 5f c4-43 a4 d0 dc c6 4a 9d 09   ....nU_.C....J..
    0020 - 00 fe e5 90 47 c4 3f 25-04 76 d3 c6 66 8d a7 8f   ....G.?%.v..f...
    0030 - 0a b3 40 16 9d 1d 5a c0-15 de da 60 b2 44 97 a9   ..@...Z....`.D..
    0040 - 0c df da 92 03 ed ee 38-13 06 e2 0a 22 bb 06 c0   .......8...."...
    0050 - 81 af 39 2d 86 93 0e 12-b5 80 e7 be dc fe cf ee   ..9-............
    0060 - da 2b 1b 26 90 a4 0f fa-f7 6a 05 7e ae f2 0b eb   .+.&.....j.~....
    0070 - 5d dc dc 11 16 c0 bf 9e-fb 69 2b 8f 66 4b 1b 85   ]........i+.fK..
    0080 - 09 59 35 4e f3 f2 27 46-df b9 bb 40 8d 1e 0b 28   .Y5N..'F...@...(
    0090 - 58 f5 f8 a8 3d a7 11 04-4b 3a df 3e 69 1a 77 df   X...=...K:.>i.w.
    00a0 - 5c a3 55 aa b6 8e 1b 25-a5 1d b0 ef 97 8a df 02   \.U....%........

    Start Time: 1616037625
    Timeout   : 7200 (sec)
    Verify return code: 0 (ok)
    Extended master secret: yes
---
closed

```

**Comment by smilexth** [comment]
> Can you give details about your mosqitto tls configuration ? Hostname ? Certificate?

```
#MQTT, NO ENCRYPTION
listener 1883
protocol mqtt

#MQTT, ENCRYPTED
listener 8883
protocol mqtt
capath /etc/ssl/certs
keyfile  /root/.caddy/acme/acme-v02.api.letsencrypt.org/sites/mqtt.xxxx.com/mqtt.xxxx.com.key
certfile /root/.caddy/acme/acme-v02.api.letsencrypt.org/sites/mqtt.xxxx.com/mqtt.xxxx.com.crt

#SMX
tls_version tlsv1.2
require_certificate false



#WEBSOCKECTS, NO ENCRYPTION
listener 8080
protocol websockets

#WEBSOCKECTS, ENCRYPTED
listener 8081
protocol websockets
capath /etc/ssl/certs
keyfile  /root/.caddy/acme/acme-v02.api.letsencrypt.org/sites/mqtt.xxxx.com/mqtt.xxxx.com.key
certfile /root/.caddy/acme/acme-v02.api.letsencrypt.org/sites/mqtt.xxxx.com/mqtt.xxxx.com.crt

#WEBSOCKETS ENCRYPTED, SAME ORIGIN ONLY
#NGINX FORWARDS PORT 443 to MOSQUITTO ON PORT 8080
#NOTE NOT ON PORT 8081!


# AUTHENTICATION PLUGINS
# There are so many plugins , requiring so much
# configuration, that they have not been tested with
# this docker container.  So please let me know if they
# work for you or not.
# Personally I am going to be using and testing the JWT plugins.
```

**Comment by Elix-g** [comment]
Your openssl output shows

`ECDHE-ECDSA-AES256-GCM-SHA384`

while the result on my machine is

`ECDHE-RSA-AES256-GCM-SHA384`.

Both ciphers are recommended for use with TLS1.2 and are defined in IETF [RFC5289](https://tools.ietf.org/html/rfc5289) and use Elliptic Curve so you can try to re-compile Tasmota with USE_MQTT_TLS_FORCE_EC_CIPHER.

**Comment by smilexth** [comment]
> ECDHE-ECDSA-AES256-GCM-SHA384

Tasmota `user_config_override.h`

![CropperCapture 6](https://user-images.githubusercontent.com/154602/111619212-e0ad2100-8817-11eb-9851-957135de8a03.png)

As your mention but i still get no luck

**Comment by Elix-g** [comment]
Can you run the nmap command as shown in my previous comment? nmap will show all ciphers supported by your broker.

**Comment by smilexth** [comment]
> Can you run the nmap command as shown in my previous comment? nmap will show all ciphers supported by your broker.


```
Starting Nmap 7.80 ( https://nmap.org ) at 2021-03-18 11:36 UTC
Nmap scan report for mqtt.xxx.com (143.198.196.xxx)
Host is up.
Other addresses for mqtt.xxx.com (not scanned): 128.199.145.xxx
rDNS record for 143.198.196.xxx: mqtt-tls

PORT     STATE    SERVICE
8883/tcp filtered secure-mqtt

Nmap done: 1 IP address (1 host up) scanned in 3.16 seconds

```

They look like no furthur information to crawling, please give me some advise

**Comment by Elix-g** [comment]
That's exactly nmap's behaviour when mosquitto is configured to use TLSv1.3. When I enable TLSv1.3 on my broker, I have the same result with nmap and Tasmota also cannot connect to the broker. So the question is, why your broker offers TLSv1.3 although you limit to TLSv1.2 by setting `tls_version tlsv1.2`. Can you rewrite your mosquitto.conf as follows?

```
listener 8883 mqtt.server.name
protocol mqtt
allow_anonymous false
tls_version tlsv1.2
cafile   /etc/mosquitto/certs/ca.crt
certfile /root/.caddy/acme/acme-v02.api.letsencrypt.org/sites/mqtt.xxxx.com/mqtt.xxxx.com.crt
keyfile  /root/.caddy/acme/acme-v02.api.letsencrypt.org/sites/mqtt.xxxx.com/mqtt.xxxx.com.key
require_certificate false
```

Pay attention to use cafile instead of capath and do not omit the hostname in the listener option.

**Comment by smilexth** [comment]
I just switch to another mosquitto tls server, It's now OK

But when i reflash with same parameter (fingerprint auto-learn) into another devices it return "Bad fingerprint" 

Do i need to generate authentication password for each devices ?
```
20:23:16.282 MQT: TLS connection error: 1
20:23:16.286 MQT: Connect failed to mqtt.xxx.com:8883, rc -2. Retry in 40 sec
```

**Comment by Elix-g** [comment, marked as answer]
Try `mqttfingerprint1 0` from the web console.

**Comment by smilexth** [comment]
> Try `mqttfingerprint1 0` from the web console.

Excellent, Thank you all guys

**Comment by DaniilMir** [comment]
Hello All.
Thanks for all your replies. 
I was playing around with Homeassistant MQTT + [DuckDNS  Letsencrypt](https://github.com/home-assistant/addons/blob/master/duckdns/README.md).
Idea was to reuse the same auto-created certificates from the WEB UI in the MQTT broker. 
It should be super easy to setup it once and have trusted certificates with auto renew. 
During the configuration I faced the same issues as topic starter had.
It was failing with MQT: TLS connection error: 296 = SSL3_ALERT_HANDSHAKE_FAILURE
I did tests with the openssl and found that self-signed certificates and Letsencrypt certificates gave me different responses:  
Letsencrypt generated automatically - ECDHE-ECDSA-AES256-GCM-SHA384 - Not working.
Self - Signed certificate - ECDHE-RSA-AES256-GCM-SHA384 - Working. 
Certificates generated with prime256v1 and secp384r1 options were not working. 
Once I added an option `algo: rsa` to the DuckDNS Letsencrypt plugin - it started working. 
response : 
```
Peer signature type: RSA-PSS
...
New, TLSv1.3, Cipher is TLS_AES_256_GCM_SHA384
Server public key is 4096 bit
```
So I have some questions and didn't want to open a new discussion for that: 

1. Is RSA-PSS secure enough to use in 2022?
2. Can we use Tasmota with Letsecnrypt CA secp384r1 instead ? I saw it in Tasmota\lib\lib_ssl\bearssl-esp8266\src\ec\ec_secp384r1.c
3. I saw options for providing TLSKey1 and TLSKey2 but they work only for AWS IOT. Goal is to set the `require_certificate: true` in the hassio MQTT broker and force all Tasmotas to provide a valid key. Can it be changed to use with Letsencrypt as well?

**Reply by s-hadinger** [reply]
1. I believe so as it is still very widely used. This only matters to authenticate your server certificate.
2. I suppose that you are talking about server certificate (ECDSA), not about the curve used for ECDH. ECDSA has a strong impact on performance last time I tried. It is also not so widely used for now. It is possible to add ECDSA and secp384r1 curves but at a cost in Flash size. There is no plan to include them in standard builds in the near future
3. Providing local keys should work as long as the server requests them and `#define USE_MQTT_AWS_IOT` is set. Only `prime256v1` private certificate are supported because of their small size.

**Reply by DaniilMir** [reply]
@s-hadinger 
Thank you for your prompt reply.
As far I understand if we are setting the `#define USE_MQTT_AWS_IOT` it will use AWS CA instead of LetsEncrypt CA check from the Tasmota side and may fail MQTT Server certificate validation.

**Reply by s-hadinger** [reply]
> setting the `#define USE_MQTT_AWS_IOT` it will use AWS CA instead of LetsEncrypt CA check from the Tasmota side

Not anymore. This was true some time ago, but now all builds contain both AWS and Letsencrypt CA certificates.

**Reply by DaniilMir** [reply]
It looks like with the letsencrypt it is not possible to create local keys.
As per [community.letsencrypt.org/](https://community.letsencrypt.org/t/can-i-create-client-certificates-for-a-received-letsencrypt-certificate/78627/3) topic - they recommend to use self signed certificates for this purpose.
But it is not clear for me - which option will be more secure. 

- Validating client keys on the MQTT broker, but having Self signed certificates.

or 

- having an MQTT server certificate generated by a well known provider, validated by tasmota device, but allowing devices without client key validation?  )

**Reply by devhammer** [reply]
> Once I added an option `algo: rsa` to the DuckDNS Letsencrypt plugin - it started working.

How long did it take to start working after you added this?

I have a Sonoff RF Bridge flashed with Tasmota, with TLS enabled. 

I'm getting the same errors on the web console (TLS connection error: 296), and the Mosquitto log is showing the error

> OpenSSL Error[0]: error:1417A0C1:SSL routines:tls_post_process_client_hello:no shared cipher

I can connect fine from MQTT-Explorer, but no luck from my Sonoff, and not sure how to troubleshoot beyond the error messages above.

I'm in the process of trying to force a renew of the cert to see if that is necessary.

**Reply by devhammer** [reply]
OK. Appears that the resolution was the same for me...just needed to force the renew of the certificate, and once HA/Mosquitto add-in was using the cert with the rsa algo, all seems well.

Also tested successfully connecting to a HiveMQ cloud instance.

And for future folks who may run across this thread...if you're running NGINX add-on in HA, you may need to restart that and DuckDNS add-on to get things to refresh.

**Reply by kracas** [reply]
I've been racking my brain and the internet for many hours trying to understand why Mosquitto + standard LetsEncrypt certificate generated with certbot was working with Shelly (Mongoose OS) and not with Sonoff (ESP32) + Tasmota. After many trials and errors I found your post. Changing the key type to RSA fixed everything. Thank you!
